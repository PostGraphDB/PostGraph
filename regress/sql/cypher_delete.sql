/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

LOAD 'postgraph';
SET search_path TO postgraph;

SELECT create_graph('cypher_delete');
CREATE GRAPH cypher_delete;
USE GRAPH cypher_delete;

--Test 1: Delete Vertices
CREATE (:v);
CREATE (:v {i: 0, j: 5, a: 0});
CREATE (:v {i: 1});

MATCH(n) DELETE n RETURN n;
MATCH(n) RETURN n;

--Test 2: Delete Edges
CREATE (:v)-[:e]->(:v);

--Should Fail
MATCH(n1)-[e]->(n2) DELETE n1 RETURN n1;
MATCH(n1)-[e]->(n2) DELETE n2 RETURN n2;

MATCH(n1)-[e]->(n2) DELETE e RETURN e;
MATCH()-[e]->() DELETE e RETURN e;

--Cleanup
MATCH(n) DELETE n RETURN n;

--Test 4: DETACH DELECT a vertex
CREATE (:v)-[:e]->(:v);
MATCH(n1)-[e]->(n2) DETACH DELETE n1 RETURN e;

--Cleanup
MATCH(n) RETURN n;

--Test 4: DETACH DELECT two vertices tied to the same edge
CREATE (:v)-[:e]->(:v);
MATCH(n1)-[e]->(n2) DETACH DELETE n1, n2 RETURN e;

--Test 4: DETACH DELECT a vertex
CREATE (:v)-[:e]->(:v);
MATCH(n1)-[e]->(n2) DETACH DELETE n1, n2 RETURN e;

--Test 5: DETACH DELETE a vertex that points to itself
CREATE (n:v)-[:e]->(n);
MATCH(n1)-[e]->() DETACH DELETE n1 RETURN e;

--Test 6: DETACH Delete a vertex twice
CREATE (n:v)-[:e]->(n);
MATCH(n1)-[e]->(n2) DETACH DELETE n1, n2 RETURN e;

--Test 7: Test the SET Clause on DELETED node
CREATE (:v);
CREATE (:v {i: 0, j: 5, a: 0});
CREATE (:v {i: 1});

MATCH(n) DELETE n SET n.lol = 'ftw' RETURN n;
MATCH(n) RETURN n;

--Test 8:
CREATE (n:v)-[:e]->(:v) CREATE (n)-[:e]->(:v);
MATCH(n1)-[]->() DELETE n1 RETURN n1;

--Cleanup
MATCH(n) DETACH DELETE n RETURN n;
MATCH(n) RETURN n;

--Test 9:
CREATE (n:v)-[:e]->(:v);
MATCH(n1)-[e]->() DELETE e, n1 RETURN n1;

--Cleanup
MATCH(n) DETACH DELETE n RETURN n;
MATCH(n) RETURN n;

--Test 10:
CREATE (n:v)-[:e]->(:v);
MATCH(n1)-[e]->() DELETE n1, e RETURN n1;

--Cleanup
MATCH(n) DETACH DELETE n RETURN n;
MATCH(n) RETURN n;

--Test 11: Delete a vertex twice
CREATE (n:v)-[:e]->(:v), (n)-[:e]->(:v) RETURN n;
MATCH(n1)-[e]->() DETACH DELETE n1 RETURN n1, e;

--Cleanup
MATCH(n) DELETE n RETURN n;
MATCH(n) RETURN n;

--Test 12:
CREATE (:v)-[:e]->(:v);
MATCH(n)-[e]->() DETACH DELETE n CREATE (n)-[:e2]->(:v) RETURN e;

--Cleanup
MATCH(n) DETACH DELETE n RETURN n;
MATCH(n) RETURN n;

--Test 13:
CREATE (n:v)-[:e]->(n);
MATCH(n)-[e]->(m) DETACH DELETE n CREATE (m)-[:e2]->(:v) RETURN e;

--Cleanup
MATCH(n) DETACH DELETE n RETURN n;
MATCH(n) RETURN n;

--Test 14:
CREATE (n:v)-[:e]->(n);
CREATE (:v)-[:e]->(:v);
MATCH(n)-[e]->(m) DETACH DELETE n CREATE (m)-[:e2]->(:v) RETURN e;

--Cleanup
MATCH(n) DETACH DELETE n RETURN n;
MATCH(n) RETURN n;

--Test 15:
CREATE (:v);
MATCH(n) SET n.i = 0 DELETE n RETURN n;

--Cleanup
MATCH(n) RETURN n;

--Test 16:
CREATE (:v);
MATCH(n) DELETE n SET n.i = 0 RETURN n;

--Cleanup
MATCH(n) RETURN n;

--Test 17:
CREATE (n:v)-[:e]->(:v);
MATCH(n)-[e]->(m) DETACH DELETE n SET e.i = 1 RETURN e;

--Cleanup
--Note: Expect 1 vertex
MATCH(n) DELETE n RETURN n;

--Test 18:
CREATE (n:v)-[:e]->(:v);
MATCH(n)-[e]->(m) SET e.i = 1 DETACH DELETE n RETURN e;

--Cleanup
MATCH(n) DELETE n RETURN n;

--Test 19:
CREATE (n:v);
MATCH (n) DELETE n CREATE (n)-[:e]->(:v) RETURN n;

--Cleanup
MATCH(n) DELETE n RETURN n;

--Test 20 Undefined Reference:
MATCH (n) DELETE m RETURN n;

--Test 21 Prepared Statements
/*
CREATE (v:v);

PREPARE d AS MATCH (v) DELETE (v) RETURN v;
EXECUTE d;

CREATE (v:v);
EXECUTE d;

--Test 22 pl/pgsql Functions
CREATE (v:v);

CREATE FUNCTION delete_test()
RETURNS TABLE(vertex vertex)
LANGUAGE plpgsql
VOLATILE
AS $BODY$
BEGIN
	RETURN QUERY MATCH (v) DELETE (v) RETURN v;
END
$BODY$;

SELECT delete_test();

CREATE (v:v);
SELECT delete_test();

-- Clean Up
MATCH(n) DETACH DELETE n RETURN n;
*/
--Test 23
CREATE (n:v)-[:e]->();
CREATE (n:v)-[:e2]->();

MATCH p=()-[]->() RETURN p;
MATCH(n)-[e]->(m)  DELETE e;
MATCH p=()-[]->() RETURN p;

-- Clean Up
MATCH(n)  DELETE n RETURN n;

--Test 24
CREATE (n:v)-[:e]->();
CREATE (n:v)-[:e2]->();

MATCH p=()-[]->() RETURN p;
MATCH(n)-[]->() DETACH DELETE n;
MATCH p=()-[]->() RETURN p;

-- Clean Up
MATCH(n) DELETE n RETURN n;

--
-- Clean up
--
DROP FUNCTION delete_test;
SELECT drop_graph('cypher_delete', true);

--
-- End
--
