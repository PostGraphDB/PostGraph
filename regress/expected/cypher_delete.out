/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
LOAD 'postgraph';
SET search_path TO postgraph;
SELECT create_graph('cypher_delete');
NOTICE:  graph "cypher_delete" has been created
 create_graph 
--------------
 
(1 row)

CREATE GRAPH cypher_delete;
ERROR:  graph "cypher_delete" already exists
USE GRAPH cypher_delete;
 use_graph 
-----------
 
(1 row)

--Test 1: Delete Vertices
CREATE (:v);
--
(0 rows)

CREATE (:v {i: 0, j: 5, a: 0});
--
(0 rows)

CREATE (:v {i: 1});
--
(0 rows)

MATCH(n) DELETE n RETURN n;
                                       n                                       
-------------------------------------------------------------------------------
 {"id": 844424930131969, "label": "v", "properties": {}}
 {"id": 844424930131970, "label": "v", "properties": {"a": 0, "i": 0, "j": 5}}
 {"id": 844424930131971, "label": "v", "properties": {"i": 1}}
(3 rows)

MATCH(n) RETURN n;
 n 
---
(0 rows)

--Test 2: Delete Edges
CREATE (:v)-[:e]->(:v);
--
(0 rows)

--Should Fail
MATCH(n1)-[e]->(n2) DELETE n1 RETURN n1;
ERROR:  Cannot delete vertex n1, because it still has edges attached. To delete this vertex, you must first delete the attached edges.
MATCH(n1)-[e]->(n2) DELETE n2 RETURN n2;
ERROR:  Cannot delete vertex n2, because it still has edges attached. To delete this vertex, you must first delete the attached edges.
MATCH(n1)-[e]->(n2) DELETE e RETURN e;
                                                        e                                                         
------------------------------------------------------------------------------------------------------------------
 {"id": 1125899906842625, "start_id": 844424930131972, "end_id": 844424930131973, "label": "e", "properties": {}}
(1 row)

MATCH()-[e]->() DELETE e RETURN e;
 e 
---
(0 rows)

--Cleanup
MATCH(n) DELETE n RETURN n;
                            n                            
---------------------------------------------------------
 {"id": 844424930131972, "label": "v", "properties": {}}
 {"id": 844424930131973, "label": "v", "properties": {}}
(2 rows)

--Test 4: DETACH DELECT a vertex
CREATE (:v)-[:e]->(:v);
--
(0 rows)

MATCH(n1)-[e]->(n2) DETACH DELETE n1 RETURN e;
                                                        e                                                         
------------------------------------------------------------------------------------------------------------------
 {"id": 1125899906842626, "start_id": 844424930131974, "end_id": 844424930131975, "label": "e", "properties": {}}
(1 row)

--Cleanup
MATCH(n) RETURN n;
                            n                            
---------------------------------------------------------
 {"id": 844424930131975, "label": "v", "properties": {}}
(1 row)

--Test 4: DETACH DELECT two vertices tied to the same edge
CREATE (:v)-[:e]->(:v);
--
(0 rows)

MATCH(n1)-[e]->(n2) DETACH DELETE n1, n2 RETURN e;
                                                        e                                                         
------------------------------------------------------------------------------------------------------------------
 {"id": 1125899906842627, "start_id": 844424930131976, "end_id": 844424930131977, "label": "e", "properties": {}}
(1 row)

--Test 4: DETACH DELECT a vertex
CREATE (:v)-[:e]->(:v);
--
(0 rows)

MATCH(n1)-[e]->(n2) DETACH DELETE n1, n2 RETURN e;
                                                        e                                                         
------------------------------------------------------------------------------------------------------------------
 {"id": 1125899906842628, "start_id": 844424930131978, "end_id": 844424930131979, "label": "e", "properties": {}}
(1 row)

--Test 5: DETACH DELETE a vertex that points to itself
CREATE (n:v)-[:e]->(n);
--
(0 rows)

MATCH(n1)-[e]->() DETACH DELETE n1 RETURN e;
                                                        e                                                         
------------------------------------------------------------------------------------------------------------------
 {"id": 1125899906842629, "start_id": 844424930131980, "end_id": 844424930131980, "label": "e", "properties": {}}
(1 row)

--Test 6: DETACH Delete a vertex twice
CREATE (n:v)-[:e]->(n);
--
(0 rows)

MATCH(n1)-[e]->(n2) DETACH DELETE n1, n2 RETURN e;
                                                        e                                                         
------------------------------------------------------------------------------------------------------------------
 {"id": 1125899906842630, "start_id": 844424930131981, "end_id": 844424930131981, "label": "e", "properties": {}}
(1 row)

--Test 7: Test the SET Clause on DELETED node
CREATE (:v);
--
(0 rows)

CREATE (:v {i: 0, j: 5, a: 0});
--
(0 rows)

CREATE (:v {i: 1});
--
(0 rows)

MATCH(n) DELETE n SET n.lol = 'ftw' RETURN n;
                                              n                                              
---------------------------------------------------------------------------------------------
 {"id": 844424930131975, "label": "v", "properties": {"lol": "ftw"}}
 {"id": 844424930131982, "label": "v", "properties": {"lol": "ftw"}}
 {"id": 844424930131983, "label": "v", "properties": {"a": 0, "i": 0, "j": 5, "lol": "ftw"}}
 {"id": 844424930131984, "label": "v", "properties": {"i": 1, "lol": "ftw"}}
(4 rows)

MATCH(n) RETURN n;
 n 
---
(0 rows)

--Test 8:
CREATE (n:v)-[:e]->(:v) CREATE (n)-[:e]->(:v);
--
(0 rows)

MATCH(n1)-[]->() DELETE n1 RETURN n1;
ERROR:  Cannot delete vertex n1, because it still has edges attached. To delete this vertex, you must first delete the attached edges.
--Cleanup
MATCH(n) DETACH DELETE n RETURN n;
                            n                            
---------------------------------------------------------
 {"id": 844424930131985, "label": "v", "properties": {}}
 {"id": 844424930131986, "label": "v", "properties": {}}
 {"id": 844424930131987, "label": "v", "properties": {}}
(3 rows)

MATCH(n) RETURN n;
 n 
---
(0 rows)

--Test 9:
CREATE (n:v)-[:e]->(:v);
--
(0 rows)

MATCH(n1)-[e]->() DELETE e, n1 RETURN n1;
                           n1                            
---------------------------------------------------------
 {"id": 844424930131988, "label": "v", "properties": {}}
(1 row)

--Cleanup
MATCH(n) DETACH DELETE n RETURN n;
                            n                            
---------------------------------------------------------
 {"id": 844424930131989, "label": "v", "properties": {}}
(1 row)

MATCH(n) RETURN n;
 n 
---
(0 rows)

--Test 10:
CREATE (n:v)-[:e]->(:v);
--
(0 rows)

MATCH(n1)-[e]->() DELETE n1, e RETURN n1;
ERROR:  Cannot delete vertex n1, because it still has edges attached. To delete this vertex, you must first delete the attached edges.
--Cleanup
MATCH(n) DETACH DELETE n RETURN n;
                            n                            
---------------------------------------------------------
 {"id": 844424930131990, "label": "v", "properties": {}}
 {"id": 844424930131991, "label": "v", "properties": {}}
(2 rows)

MATCH(n) RETURN n;
 n 
---
(0 rows)

--Test 11: Delete a vertex twice
CREATE (n:v)-[:e]->(:v), (n)-[:e]->(:v) RETURN n;
                            n                            
---------------------------------------------------------
 {"id": 844424930131992, "label": "v", "properties": {}}
(1 row)

MATCH(n1)-[e]->() DETACH DELETE n1 RETURN n1, e;
                           n1                            |                                                        e                                                         
---------------------------------------------------------+------------------------------------------------------------------------------------------------------------------
 {"id": 844424930131992, "label": "v", "properties": {}} | {"id": 1125899906842635, "start_id": 844424930131992, "end_id": 844424930131993, "label": "e", "properties": {}}
 {"id": 844424930131992, "label": "v", "properties": {}} | {"id": 1125899906842636, "start_id": 844424930131992, "end_id": 844424930131994, "label": "e", "properties": {}}
(2 rows)

--Cleanup
MATCH(n) DELETE n RETURN n;
                            n                            
---------------------------------------------------------
 {"id": 844424930131993, "label": "v", "properties": {}}
 {"id": 844424930131994, "label": "v", "properties": {}}
(2 rows)

MATCH(n) RETURN n;
 n 
---
(0 rows)

--Test 12:
CREATE (:v)-[:e]->(:v);
--
(0 rows)

MATCH(n)-[e]->() DETACH DELETE n CREATE (n)-[:e2]->(:v) RETURN e;
ERROR:  vertex assigned to variable n was deleted
--Cleanup
MATCH(n) DETACH DELETE n RETURN n;
                            n                            
---------------------------------------------------------
 {"id": 844424930131995, "label": "v", "properties": {}}
 {"id": 844424930131996, "label": "v", "properties": {}}
(2 rows)

MATCH(n) RETURN n;
 n 
---
(0 rows)

--Test 13:
CREATE (n:v)-[:e]->(n);
--
(0 rows)

MATCH(n)-[e]->(m) DETACH DELETE n CREATE (m)-[:e2]->(:v) RETURN e;
ERROR:  vertex assigned to variable m was deleted
--Cleanup
MATCH(n) DETACH DELETE n RETURN n;
                            n                            
---------------------------------------------------------
 {"id": 844424930131997, "label": "v", "properties": {}}
(1 row)

MATCH(n) RETURN n;
 n 
---
(0 rows)

--Test 14:
CREATE (n:v)-[:e]->(n);
--
(0 rows)

CREATE (:v)-[:e]->(:v);
--
(0 rows)

MATCH(n)-[e]->(m) DETACH DELETE n CREATE (m)-[:e2]->(:v) RETURN e;
ERROR:  vertex assigned to variable m was deleted
--Cleanup
MATCH(n) DETACH DELETE n RETURN n;
                            n                            
---------------------------------------------------------
 {"id": 844424930131998, "label": "v", "properties": {}}
 {"id": 844424930131999, "label": "v", "properties": {}}
 {"id": 844424930132000, "label": "v", "properties": {}}
(3 rows)

MATCH(n) RETURN n;
 n 
---
(0 rows)

--Test 15:
CREATE (:v);
--
(0 rows)

MATCH(n) SET n.i = 0 DELETE n RETURN n;
                               n                               
---------------------------------------------------------------
 {"id": 844424930132001, "label": "v", "properties": {"i": 0}}
(1 row)

--Cleanup
MATCH(n) RETURN n;
 n 
---
(0 rows)

--Test 16:
CREATE (:v);
--
(0 rows)

MATCH(n) DELETE n SET n.i = 0 RETURN n;
                               n                               
---------------------------------------------------------------
 {"id": 844424930132002, "label": "v", "properties": {"i": 0}}
(1 row)

--Cleanup
MATCH(n) RETURN n;
 n 
---
(0 rows)

--Test 17:
CREATE (n:v)-[:e]->(:v);
--
(0 rows)

MATCH(n)-[e]->(m) DETACH DELETE n SET e.i = 1 RETURN e;
                                                           e                                                            
------------------------------------------------------------------------------------------------------------------------
 {"id": 1125899906842641, "start_id": 844424930132003, "end_id": 844424930132004, "label": "e", "properties": {"i": 1}}
(1 row)

--Cleanup
--Note: Expect 1 vertex
MATCH(n) DELETE n RETURN n;
                            n                            
---------------------------------------------------------
 {"id": 844424930132004, "label": "v", "properties": {}}
(1 row)

--Test 18:
CREATE (n:v)-[:e]->(:v);
--
(0 rows)

MATCH(n)-[e]->(m) SET e.i = 1 DETACH DELETE n RETURN e;
                                                           e                                                            
------------------------------------------------------------------------------------------------------------------------
 {"id": 1125899906842642, "start_id": 844424930132005, "end_id": 844424930132006, "label": "e", "properties": {"i": 1}}
(1 row)

--Cleanup
MATCH(n) DELETE n RETURN n;
                            n                            
---------------------------------------------------------
 {"id": 844424930132006, "label": "v", "properties": {}}
(1 row)

--Test 19:
CREATE (n:v);
--
(0 rows)

MATCH (n) DELETE n CREATE (n)-[:e]->(:v) RETURN n;
ERROR:  vertex assigned to variable n was deleted
--Cleanup
MATCH(n) DELETE n RETURN n;
                            n                            
---------------------------------------------------------
 {"id": 844424930132007, "label": "v", "properties": {}}
(1 row)

--Test 20 Undefined Reference:
MATCH (n) DELETE m RETURN n;
ERROR:  undefined reference to variable m in DELETE clause
LINE 1: MATCH (n) DELETE m RETURN n;
                         ^
--Test 21 Prepared Statements
/*
CREATE (v:v);

PREPARE d AS MATCH (v) DELETE (v) RETURN v;
EXECUTE d;

CREATE (v:v);
EXECUTE d;

--Test 22 pl/pgsql Functions
CREATE (v:v);

CREATE FUNCTION delete_test()
RETURNS TABLE(vertex vertex)
LANGUAGE plpgsql
VOLATILE
AS $BODY$
BEGIN
	RETURN QUERY MATCH (v) DELETE (v) RETURN v;
END
$BODY$;

SELECT delete_test();

CREATE (v:v);
SELECT delete_test();

-- Clean Up
MATCH(n) DETACH DELETE n RETURN n;
*/
--Test 23
CREATE (n:v)-[:e]->();
--
(0 rows)

CREATE (n:v)-[:e2]->();
--
(0 rows)

MATCH p=()-[]->() RETURN p;
                                                                                                                  p                                                                                                                   
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 [{"id": 844424930132009, "label": "v", "properties": {}}, {"id": 2251799813685249, "start_id": 844424930132009, "end_id": 281474976710658, "label": "e2", "properties": {}}, {"id": 281474976710658, "label": "", "properties": {}}]
 [{"id": 844424930132008, "label": "v", "properties": {}}, {"id": 1125899906842643, "start_id": 844424930132008, "end_id": 281474976710657, "label": "e", "properties": {}}, {"id": 281474976710657, "label": "", "properties": {}}]
(2 rows)

MATCH(n)-[e]->(m)  DELETE e;
--
(0 rows)

MATCH p=()-[]->() RETURN p;
 p 
---
(0 rows)

-- Clean Up
MATCH(n)  DELETE n RETURN n;
                            n                            
---------------------------------------------------------
 {"id": 281474976710657, "label": "", "properties": {}}
 {"id": 281474976710658, "label": "", "properties": {}}
 {"id": 844424930132008, "label": "v", "properties": {}}
 {"id": 844424930132009, "label": "v", "properties": {}}
(4 rows)

--Test 24
CREATE (n:v)-[:e]->();
--
(0 rows)

CREATE (n:v)-[:e2]->();
--
(0 rows)

MATCH p=()-[]->() RETURN p;
                                                                                                                  p                                                                                                                   
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 [{"id": 844424930132011, "label": "v", "properties": {}}, {"id": 2251799813685250, "start_id": 844424930132011, "end_id": 281474976710660, "label": "e2", "properties": {}}, {"id": 281474976710660, "label": "", "properties": {}}]
 [{"id": 844424930132010, "label": "v", "properties": {}}, {"id": 1125899906842644, "start_id": 844424930132010, "end_id": 281474976710659, "label": "e", "properties": {}}, {"id": 281474976710659, "label": "", "properties": {}}]
(2 rows)

MATCH(n)-[]->() DETACH DELETE n;
--
(0 rows)

MATCH p=()-[]->() RETURN p;
 p 
---
(0 rows)

-- Clean Up
MATCH(n) DELETE n RETURN n;
                           n                            
--------------------------------------------------------
 {"id": 281474976710659, "label": "", "properties": {}}
 {"id": 281474976710660, "label": "", "properties": {}}
(2 rows)

--
-- Clean up
--
DROP FUNCTION delete_test;
ERROR:  could not find a function named "delete_test"
SELECT drop_graph('cypher_delete', true);
ERROR:  graph "ypher_delete" does not exist
--
-- End
--
